### ISSUE TITLE
Envoy crashes when upgrading WASM modules under load

### ISSUE THREAD

**Issue by klarose** [istio/istio#33091]
**Bug description**

I recently developed a c++ WASM module which manipulates response headers based on request metadata. The module seems to work fine under normal operation. However, I found that when upgrading it to a new version, every istio-proxy running it would crash pretty much in unison. The upgrade was performed by changing the uri pointing to the wasm module in the wasm EnvoyFilter. Note that I am using the wasm module distribution documented here: https://istio.io/latest/docs/ops/configuration/extensibility/wasm-module-distribution.

The issue did not reproduce if the cluster was not receiving a large amount of traffic. It appears as though something is being deleted (perhaps the wasm plugin context?) when the configuration changes, but the request instance still holds a handle to it.
I have two different backtraces. I'll paste the shorter one here, and attach the longer one.

```
2021-05-21T19:58:43.602418Z	critical	envoy backtrace	Caught Segmentation fault, suspect faulting address 0x5645000bcc60
2021-05-21T19:58:43.602462Z	critical	envoy backtrace	Backtrace (use tools/stack_decode.py to get line numbers):
2021-05-21T19:58:43.602469Z	critical	envoy backtrace	Envoy version: 66da2bf864bde982351ee0ca2cae0a4e931f923c/1.17.1/Clean/RELEASE/BoringSSL
2021-05-21T19:58:43.604067Z	critical	envoy backtrace	#0: __restore_rt [0x7fcef9aef980]
2021-05-21T19:58:43.638650Z	critical	envoy backtrace	#1: Envoy::Extensions::Common::Wasm::Context::decodeData() [0x5644fb9e074b]
2021-05-21T19:58:43.673538Z	critical	envoy backtrace	#2: Envoy::Http::FilterManager::decodeData() [0x5644fcbcdcf6]
2021-05-21T19:58:43.703375Z	critical	envoy backtrace	#3: Envoy::Http::ConnectionManagerImpl::ActiveStream::decodeData() [0x5644fcbc0da6]
2021-05-21T19:58:43.730489Z	critical	envoy backtrace	#4: Envoy::Http::Http2::ConnectionImpl::onFrameReceived() [0x5644fcbee96e]
2021-05-21T19:58:43.757338Z	critical	envoy backtrace	#5: Envoy::Http::Http2::ConnectionImpl::Http2Callbacks::Http2Callbacks()::$_18::__invoke() [0x5644fcbf5888]
2021-05-21T19:58:43.784394Z	critical	envoy backtrace	#6: nghttp2_session_on_data_received [0x5644fcd4a5ac]
2021-05-21T19:58:44.198038Z	error	Epoch 0 exited with error: signal: segmentation fault (core dumped)
```
[envoy-crash.txt](https://github.com/istio/istio/files/6542437/envoy-crash.txt)

[ ] Docs
[ ] Installation
[ X ] Networking
[ ] Performance and Scalability
[ ] Extensions and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure
[ ] Upgrade

**Expected behavior**

New WASM plugin loaded without any request loss. It is acceptable for existing requests to continue to run against the old plugin.

**Steps to reproduce the bug**

1. Configure istio to use the wasm module distribution (https://istio.io/latest/docs/ops/configuration/extensibility/wasm-module-distribution/) to load a wasm module.
2. Send requests at a high rate (e.g. `while [ 1 ]; do curl https://my-site.example.com/; done`).
3. Change the wasm module to point to a different version.
4. Watch proxy logs. See it crash and restart. See all in-flight requests drop.

**Version** (include the output of `istioctl version --remote` and `kubectl version --short` and `helm version --short` if you used Helm):

In my GKE env:

```
istioctl version --remote
client version: 1.9.5
control plane version: 1.9.5
data plane version: 1.9.5 (50 proxies)
```

In my microk8s env:

```
istioctl version
client version: 1.9.5
control plane version: 1.9.5
data plane version: 1.9.5 (21 proxies)
```

**How was Istio installed?**
Used istioctl install -f <config-file>

**Environment where the bug was observed (cloud vendor, OS, etc)**

 - GKE
 - independently, microk8s on Ubuntu 20.04

**Comment by bianpengyuan** [comment]
cc @PiotrSikora 

It would be really helpful if you could provide a minimal wasm module code snippet which can reproduce the issue.

**Comment by PiotrSikora** [comment]
Could you enable logging at the trace level? See: https://knative.tips/troubleshooting/debugging-envoy/

**Comment by klarose** [comment]
@bianpengyuan @PiotrSikora 

I'll get that info shortly.

**Comment by klarose** [comment]
Took longer to get the minimal code than I expected. You can find two module here: https://github.com/klarose/envoy-wasm-tests. 'nop' and 'static-header', underneath the extensions folder. I have config files in each which you can use to load. You can also find a config file under other-config/basic-auth which loads the istio-ecosystem basic-auth module.

Swapping between static-header and nop doesn't seem to cause the problem consistently (possibly not ever... I'm not sure I trust some of my other tests). Swapping between static-header and basic-auth does seem to consistently cause the problem.

To be clear, you can swap by doing:

```
kubectl apply -f extensions/static-header/config/static-header.yaml
<wait for some time so the extensions load>
kubectl apply -f other-config/basic-auth.yaml
```

I'll look into getting the trace output now.

**Comment by klarose** [comment]
@PiotrSikora  here's the last second or so of logs before istio crashed. In this case I didn't have as much traffic flowing through the system -- basically one long lived request at a time (I modified the upstream to delay before responding).

[istio-stripped.log](https://github.com/istio/istio/files/6548693/istio-stripped.log)

In this case, it looks like the stack was corrupted, unfortunately. Let me know if you want me to try again.

**Comment by PiotrSikora** [comment]
@klarose thanks for the reproducers, but I cannot find `other-config/basic-auth` in that repo. Did you commit it? Also, we need logs at the `trace` level (`level=trace`), `debug` isn't sufficient.

But if you're changing `.wasm`, then it seems that the Wasm filter is ejected from Envoy while still serving traffic. @kyessenov, are we using ECDS already by default? cc @mathetake for visibility

**Comment by klarose** [comment]
@PiotrSikora whoops, sorry about that. Forgot to push. It's there now.

I'll rerun with the trace level. I foolishly copied the command line in the example you gave without actually looking closely at it. Sorry!

**Comment by klarose** [comment]
@PiotrSikora Here you go. The backtrace is a little nicer this time too.

[istio.trace.stripped.log](https://github.com/istio/istio/files/6549310/istio.trace.stripped.log)

**Comment by kyessenov** [comment]
@PiotrSikora yes we use ECDS for Wasm by default. That's how URL interception works.

**Comment by mathetake** [comment]
FWIW I just spent a couple hours trying to reproduce this only with raw Envoy configs (both with/out ADS) and `tools/extensionserver/main/main.go` for variety of Wasm binaries both located at both remote/local, but I couldn't reproduce with the HEAD of Envoy (https://github.com/envoyproxy/envoy/commit/a7fa3311ad1f847d80e139970d5325a4efcbdbc1).. so I guess this is Istio specific..

**Comment by klarose** [comment]
@mathetake Do you have any thoughts on whether these crashes could be caused by the modules themselves -- e.g. if my compilation process is bad? I'm not sure about what sort of protections the wasm sandbox provides. For example, if I've somehow broken the 'destructor' interface of the module, could that be part of the issue? 

Or, from the trace I pasted above, is it more likely that the filter chain is hanging on to a dangling pointer?

**Comment by bianpengyuan** [comment]
> FWIW I just spent a couple hours trying to reproduce this only with raw Envoy configs (both with/out ADS) and `tools/extensionserver/main/main.go` for variety of Wasm binaries both located at both remote/local, but I couldn't reproduce with istio/proxyv2:1.9.5 the HEAD of Envoy ([envoyproxy/envoy@a7fa331](https://github.com/envoyproxy/envoy/commit/a7fa3311ad1f847d80e139970d5325a4efcbdbc1)).. so I guess this is Istio specific..

@mathetake with https://istio.io/latest/docs/ops/configuration/extensibility/wasm-module-distribution wasm modules should all be loaded as local files. Curious, did you test with many worker threads, and with the exact configuration @klarose provides (i.e. both old and new filter do not have vm id or root id)? Those could make differences when triggering subtle object lifecycle issue..

**Comment by mathetake** [comment]
I did with vm_id and configurations, so let me try again without vm_id and configuration. Also let me try @klarose's code (I only did with Go SDK).

**Comment by mathetake** [comment]
Update: succeeded to reproduce locally without Istio and for both Envoy's HEAD and istio/proxy, but couldn't identify the root cause.. stay tuned.

@klarose I'm pretty confident that this is an Envoy specific issue and not of Wasm/VM/Compiler's.

**Comment by mathetake** [comment]
OK, identified the root cause (which resides in Proxy-Wasm C++ Host). Will submit a fix within a few days.

**Comment by klarose** [comment]
> OK, identified the root cause (which resides in Proxy-Wasm C++ Host). Will submit a fix within a few days.

Nice! Thanks!

**Comment by mathetake** [comment]
update: actually turns out that this is an issue around Envoy's plugin lifecycle vs stream filters' life cycle, so I fixed https://github.com/envoyproxy/envoy/pull/16795 but I faced another lifecycle issue, and am working on it now.
